io/src/csv/read_cells_alt.cpp

```diff
@@ -0,0 +1,190 @@
+                std::to_string(result.module));
+        }
+
+        // Set the value on the module description.
+        const auto& binning_data = geo_it->segmentation.binningData();
+        assert(binning_data.size() >= 2);
+        result.pixel = {binning_data[0].min, binning_data[1].min,
+                        binning_data[0].step, binning_data[1].step};
+    }
+
+    // Initialize the result module's limits to the cell's properties
+    result.range0[0] = c.channel0;
+    result.range0[1] = c.channel0;
+    result.range1[0] = c.channel1;
+    result.range1[0] = c.channel1;
+
+    return result;
+}
+
+alt_cell_collection_types::host read_cells_alt(
+    std::string_view filename, const geometry* geom,
+    const digitization_config* dconfig, vecmem::memory_resource* mr) {
+
+    // Construct the cell reader object.
+    auto reader = make_cell_reader(filename);
+
+    // Create cell counter vector.
+    std::vector<cell_counter> cellCounts;
+    cellCounts.reserve(5000);
+
+    // Create a cell collection, which holds on to a flat list of all the cells
+    // and the position of their respective cell counter.
+    std::vector<std::pair<csv::cell, std::size_t>> allCells;
+    allCells.reserve(50000);
+
+    // Read all cells from input file.
+    csv::cell iocell;
+    while (reader.read(iocell)) {
+
+        // Look for current module in cell counter vector.
+        auto rit =
+            std::find_if(cellCounts.rbegin(), cellCounts.rend(),
+                         [&iocell](const cell_counter& cc) {
+                             return cc.module.module == iocell.geometry_id;
+                         });
+        // Add new cell and new cell counter if a new module is found
+        if (rit == cellCounts.rend()) {
+            const cell_module mod = get_module(iocell, geom, dconfig);
+            allCells.push_back({iocell, cellCounts.size()});
+            cellCounts.push_back({mod, 1});
+            // Add a new cell and update cell counter if repeat module is found
+        } else {
+            allCells.push_back(
+                {iocell, std::distance(cellCounts.begin(), rit.base()) - 1});
+            ++(rit->nCells);
+            rit->module.range0[0] =
+                std::min(rit->module.range0[0], iocell.channel0);
+            rit->module.range0[1] =
+                std::max(rit->module.range0[1], iocell.channel0);
+            rit->module.range1[0] =
+                std::min(rit->module.range1[0], iocell.channel1);
+            rit->module.range1[0] =
+                std::max(rit->module.range1[1], iocell.channel1);
+        }
+    }
+
+    // Transform the cellCounts.nCells vector into a prefix sum for accessing
+    // positions in the result vector.
+    std::partial_sum(cellCounts.begin(), cellCounts.end(), cellCounts.begin(),
+                     [](const cell_counter& lhs, cell_counter& rhs) {
+                         rhs.nCells += lhs.nCells;
+                         return rhs;
+                     });
```

Comment:
Maybe I am misunderstanding, but why is this necessary? If we are reading everything into a flat array there is no need to calculate the positions in the result vector, right? If we simply read the cells into memory in the same order as they appear in the file there should be no issue, I think.

Reply:
The problem is the sorting.  The algorithm relies on the fact that the cells are sorted by both module and channel1. The purpose of this is to not rely on the input file already being sorted.

Reply:
If we do assume input data is already sorted by module that would make this all very simple indeed. Even more so if we also assume they're sorted by channel1 as well.

Reply:
I think it's fair to make that assumption. In the end, we can always check for this sorted-ness and sort the data (on the GPU) if necessary.

Reply:
@krasznaa would be good to have your two cents on this. iirc we talked about this when a sometime ago and you didn't seemed very willing to depend on this assumption.

Reply:
I ended up leaving the (overcomplicated) code to allow input data to not come in sorted by module, since we're currently not too concerned about file reading times and it is safer as is.